metals/src/main/scala/scala/meta/internal/worksheets/WorksheetProvider.scala

```diff
@@ -0,0 +1,389 @@
+      }
+    }
+    statement.out.linesIterator.foreach { line =>
+      out.append("\n// ").append(line)
+    }
+    out.toString()
+  }
+
+  private def renderContentText(
+      statement: Statement,
+      margin: Int,
+      isEmptyValue: Boolean
+  ): String = {
+    val out = new StringBuilder()
+    out.append(commentHeader)
+    if (isEmptyValue) {
+      if (!statement.out.isEmpty()) {
+        out.append(statement.out.linesIterator.next())
+      }
+    } else {
+      val isSingle = statement.binders.lengthCompare(1) == 0
+      statement.binders.iterator.zipWithIndex.foreach {
+        case (binder, i) =>
+          if (!isSingle) {
+            out
+              .append(if (i == 0) "" else ", ")
+              .append(binder.name)
+              .append("=")
+          }
+          val truncatedLine = BlackWhite
+            .tokenize(binder.value, width = margin)
+            .map(_.getChars)
+            .filterNot(_.iterator.forall(_.isWhitespace))
+            .flatMap(_.iterator)
+            .filter {
+              case '\n' => false
+              case _ => true
+            }
+            .take(margin)
+          out.appendAll(truncatedLine)
+      }
+    }
+    out.toString()
+  }
+
+  private def evaluateWorksheet(
+      path: AbsolutePath,
+      token: CancelToken
+  ): Array[DecorationOptions] = {
+    val input = path.toInputFromBuffers(buffers)
+    // NOTE(olafurpg): the sbt dialects is the closest available syntax to worksheets.
+    val WorksheetDialect = dialects.Sbt1
+    val decorations = for {
+      ctx <- getContext(path)
+      source <- WorksheetDialect(input).parse[Source].toOption
+    } yield renderDecorations(path, ctx, source)
+    decorations.getOrElse(Array.empty)
+  }
+
+  private def renderDecorations(
+      path: AbsolutePath,
+      ctx: Context,
+      source: Source
+  ): Array[DecorationOptions] = {
+    val sectionInput = SectionInput(
+      path.toInputFromBuffers(buffers),
+      source,
+      Modifier.Default()
+    )
+    val sectionInputs = List(sectionInput)
+    val instrumented = Instrumenter.instrument(sectionInputs)
+    val rendered = MarkdownCompiler.buildDocument(
+      ctx.compiler,
+      ctx.reporter,
+      sectionInputs,
+      instrumented,
+      path.toString
+    )
+
+    val decorations = for {
+      section <- rendered.sections.iterator
+      statement <- section.section.statements
+    } yield renderDecoration(statement)
+
+    diagnostics.onPublishDiagnostics(
+      path,
+      reporter.diagnostics.toSeq,
+      isReset = true
+    )
+
+    decorations
+      .filterNot(_.renderOptions.after.contentText == commentHeader)
+      .toArray
+  }
+
+  private def renderDecoration(statement: Statement): DecorationOptions = {
+    val pos = statement.position
+    val range = new l.Range(
+      new l.Position(pos.startLine, pos.startColumn),
+      new l.Position(pos.endLine, pos.endColumn)
+    )
+    val margin = math.max(
+      20,
```

Comment:
Move `20` to a value? Seems like a magic number

Reply:
Good idea, done